compiler/src/dotty/tools/dotc/core/TypeComparer.scala

```diff
@@ -3095,7 +3095,7 @@ class TrackingTypeComparer(initctx: Context) extends TypeComparer(initctx) {
     def paramInstances(canApprox: Boolean) = new TypeAccumulator[Array[Type]]:
       def apply(insts: Array[Type], t: Type) = t match
         case param @ TypeParamRef(b, n) if b eq caseLambda =>
-          def range1(tp: Type) = Range(tp, tp)
+          def range1(tp: Type) = if variance == 0 then tp else Range(tp, tp)
```

Comment:
I think that's still not as precise as the original solution. The original solution says:

> Don't widen if the compared-to type contains variant named type parameters

Whereas here we implement a global criterion: 

> Under mode `canWidenAbstract = true`, don't instantiate variant named type parameters.

The original proposal would allow the instantiation of variant type parameters if they appear in a part of the pattern that does not correspond to a widened part of the scrutinee. 

It would be really good to work out the theory here. The fact that this is missing in the match type papers and thesis is a big problem.